### PR TITLE
ENH : add function to add displayhook

### PR DESCRIPTION
**Pull request by tacaswell**
This is still a prototype, but it seems to work. 

- [x] make figures be able to keep track of if they need to be re-drawn.  There is already a call back system  in-place for when `Artist` properties change (see `Artist.{add,remove}_callback` and `Artist.pchanged`) which we can tap into to set axes/figure trees as 'dirty'.  This will useful if we ever want to start caching rasters of each artist (does that even make sense?  I think it does and is an obvious place to get major gains if we do not re-render the text each time)
- [x] use the plugable interface in IPython
- [x] make sure that calling the install function multiple times does not do funny things.
- [x] make sure mplot3d is not broken

calling

```python
plt.install_repl_hook()
fig, ax = plt.subplots()
ax.plot(range(5))
```
will automatically re-draw the figure as expected.

If we are happy with this, running the installer should probably be part of the pyplot import.

attn @efiring this is the alternative to the approach in #4082 that I was talking about in MD.

### CODE REVIEW

**Comment by efiring** [comment]
@tacaswell, Thanks, now I understand.  It works on the normal ipython console, but not with the ipython qtconsole or the notebook.  

In the qtconsole:
```
In [1]: import sys

In [2]: sys.displayhook
Out[2]: <IPython.kernel.zmq.displayhook.ZMQShellDisplayHook at 0x7f01c04edda0>
```
If the present IPython API allows us to tap into that, then this seems like an elegant way to go.

**Comment by tacaswell** [comment]
With which backend? I would not expect this to work with inline as you don't have a live figure.

**Comment by efiring** [comment]
@tacaswell, I don't think I understand your last question.  Running in a notebook or qtconsole with %matplotlib, one does have a live figure.  Using qtconsole in particular is just like using a normal console.   Somehow, however, the attempt to replace the ZMQShellDisplayHook is doing nothing:
```
IPython QtConsole 3.0.0-b1
Python 3.4.0 (default, Apr 11 2014, 13:05:11) 
Type "copyright", "credits" or "license" for more information.

IPython 3.0.0-b1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
%guiref   -> A brief reference about the graphical user interface.

In [1]: %matplotlib
Using matplotlib backend: Qt4Agg

In [2]: import matplotlib.pyplot as plt

In [3]: import sys

In [4]: sys.displayhook
Out[4]: <IPython.kernel.zmq.displayhook.ZMQShellDisplayHook at 0x7f115dc47eb8>

In [5]: plt.install_repl_displayhook()

In [6]: sys.displayhook
Out[6]: <IPython.kernel.zmq.displayhook.ZMQShellDisplayHook at 0x7f115dc47eb8>
```

**Comment by tacaswell** [comment]
Ah, i now understand.  I assume this is something funny going on with where things get executed?

I threw some print statements in and

```python
Python 3.4.2 (default, Jan 20 2015, 14:24:24) 
Type "copyright", "credits" or "license" for more information.

IPython 2.3.1 -- An enhanced Interactive Python.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
%guiref   -> A brief reference about the graphical user interface.
Using matplotlib backend: Qt4Agg

In [1]: import matplotlib.pyplot as plt

In [3]: import sys

In [4]: sys.displayhook
Out[4]: <IPython.kernel.zmq.displayhook.ZMQShellDisplayHook at 0x7ff1d17e8be0>

In [5]: plt.install_repl_displayhook
Out[5]: <function matplotlib.pyplot.install_repl_displayhook>

In [6]: plt.install_repl_displayhook()
<IPython.kernel.zmq.displayhook.ZMQShellDisplayHook object at 0x7ff1d17e8be0>
<function install_repl_displayhook.<locals>.displayhook at 0x7ff1c804c6a8>


In [8]: sys.displayhook
Out[8]: <IPython.kernel.zmq.displayhook.ZMQShellDisplayHook at 0x7ff1d17e8be0>

```

(missing input/output is where I removed errors because I can't type)

I was skimming the IPython code last night and kept seeing calls to update the user namespace which I suspect is related to this.

I'll see if reviving the monkey-patching version of this work...

**Comment by tacaswell** [comment]
@minrk Thoughts on this? 

What are we doing wrong and how bad of an idea is the monkey patching?

**Comment by minrk** [comment]
@tacaswell IPython's display machinery is pluggable, it's just not on the DisplayHook object. If you want to call `draw_all()` at the end of execution in IPython, the place to do it is on the post_execute event, not displayhook:

```python
from IPython import get_ipython
ip = get_ipython()
# IPython >= 2
try:
    ip.events.register('post_execute', draw_all)
except AttributeError:
   # IPython 1.x
   ip.register_post_execute(draw_all)
```

This is what the `%matplotlib inline` magic does \*

\* There are actually two events - `post_execute` and `post_run_cell`. `post_execute` fires on *any* code execution, even non-interactive executions (side effects of UI elements), whereas `post_run_cell` only fires on the subset of executions that are "a user typed some code and ran it," which maps onto when displayhook is actually called.

**Comment by tacaswell** [comment]
last commit from right before running out of the office, not tested yet.

**Comment by efiring** [comment]
It is missing something.  With plain ipython and %matplotlib, the drawing is not triggered until the window is resized, or a zoom/pan is executed.

**Comment by tacaswell** [comment]
@efiring I can't reproduce this:

```python
(dd_py3k)[tcaswell@arya ~]$ ipython
Python 3.4.2 |Continuum Analytics, Inc.| (default, Oct 21 2014, 17:16:37) 
Type "copyright", "credits" or "license" for more information.

IPython 2.3.1 -- An enhanced Interactive Python.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import matplotlib.pyplot

In [2]: matplotlib.pyplot.install_repl_displayhook()

In [3]: %matplotlib
Using matplotlib backend: TkAgg

In [4]: import matplotlib.pyplot as plt

In [5]: ax = plt.gca()

In [6]: ax.plot(range(3))
Out[6]: [<matplotlib.lines.Line2D at 0x7f13740f0710>]

```

I will add a commit installing the hook by default as you seem happy with this approach.

**Comment by tacaswell** [comment]
This will work well for simple plots, but I think will fall on it's face for people with complex expensive (like seconds) draw time plots.

We either need to add a way to control this (rc param that `draw_all` looks at, a replhook-uninstall method, both) or to make figures smart enough to tell if they need to be re-drawn.

We probably should do both but do the switch first as it is easier/faster so we can get this merged and get people to play with it.

**Comment by tacaswell** [comment]
current failure is related to sphinx 1.3 issues

**Comment by efiring** [comment]
As an example of the problem with redrawing a complex plot when it shouldn't, try this in ipython,
typing one line at a time:
```python
%matplotlib
import matplotlib.pyplot as plt
fig, axs = plt.subplots(20, 20)
5*10
```
First, creating the 400 empty subplots is horribly slow; this has been pointed out before, and remains an area where we need some streamlining.  Redrawing is not as slow, but it still leads to a very noticeable lag in getting the prompt back after the trivial operation on the last line.

**Comment by tacaswell** [comment]
@efiring I think I have (to first order) fixed that problem and started to make the `Artists` play nice with this as well.  Try

```python
%matplotlib
import matplotlib.pyplot as plt

fig, axs = plt.subplots(2, 2)
ln, = axs[0, 0].plot(range(3))
ln.set_color('r')
ln.set_linewidth(5)
```

line-by-line.  Each of the 'set_*' calls should auto-magically redraw.

**Comment by tacaswell** [comment]
The exact plumbing of the call-backs probably needs some work.  It may be a better idea to allow the artists to mark them selves as not-dirty in their `draw` functions rather than doing so in the `Axes`/`Figure` `draw` methods. 

This is where something like traitlets comes in super handy as we could stick the 'mark as dirty' logic it to some meta-class/class level feature and have it centralized (instead of spread out across all of the `set_*` methods of every `Artist` sub-class).

**Comment by efiring** [comment]
Each draw method does seem to be a logical place to put the marking.  (Maybe we could call things "stale" rather than "dirty"--somehow it sounds better, and seems a little more descriptive of the actual state.)

**Comment by tacaswell** [comment]
I like stale better too, dirty feels too negative.

There are 74 draw methods, doing it this way wad more expedient to make
sure that this was the right route to head down.

**Comment by tacaswell** [comment]
@efiring Any thoughts on how this is coming along?   

I regretting not sorting out how to test this ahead of time.

**Comment by tacaswell** [comment]
@efiring I think I am done with the code part of this!  Now it just needs documentation, some tests, and a whole lot of documentation.

**Comment by efiring** [comment]
I think that much of the testing will have to be manual.  I will try to take a closer look some time in the next few days.

**Comment by tacaswell** [comment]
@matplotlib/developers Can people take a look at this?  I would like to get this merged and beaten upon as soon as possible.

**Comment by efiring** [comment]
@tacaswell, are the `draw_if_interactive` calls in `pyplot` functions still needed?  I would think they could all be eliminated.

**Comment by jenshnielsen** [comment]
I have had a quick test of it and it seems to be a great improvement to interactive work. 

I guess it would make sense to at least add a warning to draw_if_interactive essentially stating that it is no longer needed?

File: doc/users/whats_new/2015-05_interactive_OO.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+   ln, = ax.plot([0, 1, 4, 9, 16])
+
+   plt.show()
+
+   ln.set_color('g')
+
+
+will automatically update the plot to be green.  Subsequent
+modifications to the `Artist` objects will do likewise.
```
Comment (WeatherGod) [inline doc/users/whats_new/2015-05_interactive_OO.rst]: This might need to be made a bit clearer. I think you mean for any situation where plt.ion() is effectively on. In normal default modes, even in a REPL (but not ipython), execution stops at plt.show() until the figure window is closed. So, I would imagine many people would be confused by this example.

File: lib/matplotlib/contour.py [hunk 2]
```diff
@@ -964,7 +964,7 @@ def __init__(self, ax, *args, **kwargs):
                     segs,
                     antialiaseds=aa,
                     linewidths=width,
-                    linestyle=[lstyle],
+                    linestyles=[lstyle],
```
Comment (WeatherGod) [inline lib/matplotlib/contour.py]: unrelated change?

File: lib/matplotlib/lines.py [hunk 3]
```diff
@@ -898,6 +925,8 @@ def set_color(self, color):
 
         ACCEPTS: any matplotlib color
         """
+        if color != self._color:
```
Comment (WeatherGod) [inline lib/matplotlib/lines.py]: need to watch out here, given that there can be multiple ways to represent the same color.

File: lib/matplotlib/contour.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/contour.py]: This is from https://github.com/tacaswell/matplotlib/commit/154cd87cca035ef040f2370709726ccd508f260c and is intentional so that the kwarg is handled by the named keyword in the `__init__` signature rather than falling through the kwargs to the `set_*` magic. 

This should go in either way but I can pull it out of this PR if you want.

File: lib/matplotlib/lines.py (hunk 3 above)
Comment (tacaswell) [inline lib/matplotlib/lines.py]: When I started this I was trying to be clever and short-circuit the stale logic in cases where there is no change and then gave up an that made made calling `set_*` to always mark the artist (and hence up the chain) as stale.

In this case, even if they are the same color with different reps, marking as stale is not _really_ a problem.

File: lib/matplotlib/offsetbox.py [hunk 4]
```diff
@@ -1205,6 +1225,7 @@ def __init__(self, arr,
                  **kwargs
                  ):
 
+        OffsetBox.__init__(self)
```
Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: what is the rationale for moving this init?

File: lib/matplotlib/offsetbox.py [hunk 5]
```diff
@@ -1330,6 +1352,13 @@ def __init__(self, offsetbox, xy,
 
         other parameters are identical to that of Annotation.
         """
+
+        martist.Artist.__init__(self, **kwargs)
+        _AnnotationBase.__init__(self,
+                                 xy,
+                                 xycoords=xycoords,
+                                 annotation_clip=annotation_clip)
+
```
Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: same question here, why is it moved?

File: lib/matplotlib/offsetbox.py (hunk 4 above)
Comment (tacaswell) [inline lib/matplotlib/offsetbox.py]: Because the `self._stale` attribute is added in the parent classes` __init__` and some of the `set_*` methods used in this `__init__` require it to be there so the parent class must be inited first.

File: lib/matplotlib/contour.py (hunk 2 above)
Comment (WeatherGod) [inline lib/matplotlib/contour.py]: nah, that makes sense.

File: lib/matplotlib/lines.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/lines.py]: true. it is no worse than before

File: lib/matplotlib/offsetbox.py (hunk 4 above)
Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: looking back at the code: more precisely, the Artist class defines the `stale` property, and so if you want to set a value to that property during your constructor, then you need to construct your parent classes first. So these changes make sense. I just hope we didn't have code anywhere that depended on the older ordering

File: lib/matplotlib/lines.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/lines.py]: As for the logic to check for change in values... I could imagine a magical decorator that would examine the name of the decorated function, figure out the property it is setting, and then automatically add the logic for comparing the original value with the new value and setting the property. Although that would get a bit tricky for any setters that obtains values from the rcParams, or any other sort of normalization.

Also might be tricky to apply everywhere if the setters are used as the way to initialize the attributes (so you can't access it prior to setting it).

File: lib/matplotlib/offsetbox.py (hunk 4 above)
Comment (tacaswell) [inline lib/matplotlib/offsetbox.py]: I found most of these issues due to examples failing to build.  I do not recall seeing anything in the `__init__` methods that are doing subtle things like expecting the child classes to have pre-set some value, but silently use a default value if they have not.

My suspicion is that the current ordering was not by design.

Comment (WeatherGod) [inline lib/matplotlib/offsetbox.py]: Indeed, I think that is the case in matplotlib proper. I know that in mplot3d, I had to be a little careful with construction, though. Given that the mplot3d tests did not fail, though, I am going to assume that all is well there (although it would make sense to double-check that interactivity for mplot3d isn't impacted).

Comment (tacaswell) [inline lib/matplotlib/offsetbox.py]: As I recall the terrifying thing in mplot3d is the class promotion which I don't think this will have an impact on.

**Comment by tacaswell** [comment]
@ChrisBeaumont @astrofrog These changes may be of interest for you and glue.  As I recall, glue is doing some terrifying monkey patching to manage deferred draws, I think this might provide an alternative 'native' method to do the same.   

I would also be interested in use cases that the monkey patching supports which this does not.

**Comment by ChrisBeaumont** [comment]
@tcaswell thanks for the heads up! Yeah this looks like a similar problem to what we try to solve in Glue. 

Right now our approach is to eagerly call draw within any function that changes plot state, and then use a context manager (+ decorator) to prevent multiple renders within a single context block (it also has the important property that nested context managers skip renders entirely, so that draw is executed exactly once in at the end of some event cascade)

It looks like the approach recommend by this PR is to stop calling draw eagerly at state-change time, and either:

1) Refactor our context manager to check if the figure is stale and, if so, redraw
2) Do away with the context manager entirely, and add a callback akin to `draw_all` to some hook (it's less obvious to me what this hook is in a non-REPL environment. Maybe we could use Qt's heartbeat or processEvents function, but I need to read up on how Qt schedules the handoff between Qt and Python. We want to avoid both overly frequent or infrequent calls to draw_all)

**Comment by tacaswell** [comment]
@efiring I don't see why hold requires `draw_if_interactive`.

**Comment by tacaswell** [comment]
@WeatherGod mplot3D seems to work just fine.  I also added the stale flag logic a couple of places.

**Comment by tacaswell** [comment]
well, it worked fine until I broke it.  I will fix it tomorrow.

**Comment by efiring** [comment]
Regarding "hold" (which I would be delighted to dump--it seems like a useless holdover from Matlab), the point is that the hold state is saved, it is replaced with the hold kwarg value, then the draw occurs, and then the hold state is restored.  I don't see how to do this simply if the draw occurs only when the displayhook is called.  That will be *after* the pyplot function has executed.  It seems like it would require a second hold flag which would be set based on the kwarg.

**Comment by tacaswell** [comment]
`hold` is not used in the `draw` logic, it is used in the plotting calls to determine if we need to call `cla` or not.  I think we could re-order the boiler plate code to

```pyhon
washold = ax.ishold()
if hold is not None
    ax.hold(hold)
try:
    ret = ax.foo(..)
finally:
    ax.hold(washold
draw_if_interactive()
```
with no external change.

I am also strongly in favor of ditching `hold`   If we don't get rid of hold altogether I would like to push all of the logic into `pyplot` and out of the `Axes` class.  If we can't do that, I would like to at least turn this into a context, but that is another PR all together.

**Comment by efiring** [comment]
That looks right; the cla() is called or not in `ax.foo(..)` depending on `ax.ishold()`.  Is there any reason to leave in the `draw_if_interactive()`?  Wouldn't the hook cause the draw to be executed here anyway?

File: doc/users/whats_new/2015-05_interactive_OO.rst [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+To facilitate interactive usage a ``draw_all`` method has been added
+to ``pyplot`` which will redraw all of the figures which are 'stale'.
+
+To make this convenient for interactive use matplotlib now registers
+a function either with IPython's 'post_execute' event or with the
+displayhook in the standard python REPL to automatically call
+``plt.draw_all`` just before control is returned to the REPL.  This ensures
+that the draw command is deferred and only called once.
+
+The upshot of this is that for interactive backends (including notebook)
```
Comment (jenshnielsen) [inline doc/users/whats_new/2015-05_interactive_OO.rst]: I think we should make it a bit more explicit to make sure that users aren't confused between the notebook with inline backend and the notebook backend. I would write something like "(including notebook i.e. '%matplotlib notebook') "

**Comment by tacaswell** [comment]
I looked at removing `draw_if_interactive` and discovered I missed parts of `Axes` and all of `Axis`.  It is partially done in my computer but have not pushed yet (and want to take another look through before I do).

**Comment by tacaswell** [comment]
Removed `draw_if_interactive`, but it still needs testing + documentation.

One corner case where `draw_if_interactive` might do a better is with backends that are not interactive. even if the user has set `plt.ion` (ex, the png backend).

**Comment by tacaswell** [comment]
Re-based this again.

@efiring @WeatherGod @jenshnielsen I think I have addressed the documentation issues.  I would like to get this merged as soon as possible to avoid excessive re-bases and to get all the brave souls who run master testing this :smiling_imp: .

**Comment by efiring** [comment]
Looks good; any additional wrinkles can be ironed out in subsequent PRs.  This is a huge step forward.

**Comment by mdehoon** [comment]
@tacaswell , @efiring  A few comments:
>Having artists be aware of the canvas anywhere out side of draw does a good deal of damage to the separation between the front and backend layers.

I think this is a misunderstanding. Calling canvas.draw_idle doesn't actually draw anything; it just tells the canvas to redraw itself on the next pass through the event loop. This is standard practice in GUI design and does not break the separation between the frontend and the backend.

> If I understand what you are suggesting correctly, you want to replace all of the self.stale = True lines with
```python
if self.ax:
    if self.ax.figure:
        if self.ax.figure.canvas:
            if matplotlib.is_interactive():
                self.ax.figure.canvas.draw_idle()
```

Something like that, yes, but it can probably be simplified, and the call to ```matplotlib.is_interactive``` would probably go somewhere else.

@efiring 
> it is simply not possible to support full interactivity without either using IPython, which has a 'post-execute' hook,

Plain python also has a post-execute hook, which is PyOS_InputHook; this is what the GUI backends use to be interactive. The call to ```draw_idle``` lets you make use of this hook.

**Comment by efiring** [comment]
On 2015/07/11 7:33 PM, mdehoon wrote:
> Plain python also has a post-execute hook, which is PyOS_InputHook; this
> is what the GUI backends use to be interactive. The call to |draw_idle|
> lets you make use of this hook.
>

Right, but PyOS_InputHook is at the C level; the only hook at the Python 
level is the displayhook, which doesn't get called when we need it. 
Ipython uses their own REPL which provides the Python-level post-execute 
hook.

The Artist picker already accesses figure.canvas, so I thought maybe the 
stale setter could also call figure.canvas.draw_idle() with a try/except 
to catch AttributeError.  It's not that simple, though.  It's easy to 
land in an infinite recursion.  I haven't figured it out yet.

**Comment by efiring** [comment]
@tacaswell, in the process of unsuccessfully working on this, I put in a counter to see how many times the `stale` setter is called for the single plot call script that @mdehoon provided here for illustration.  The total is 3529. That's a bit frightening, though perhaps it shouldn't be.

**Comment by efiring** [comment]
@tacaswell,  Now I see your comment from 4 hours ago--I missed it originally.  And I forgot that there was still one pending PR in this series, but had only the vague recollection that everything had been worked out.  Good!

**Comment by mdehoon** [comment]
@efiring 
>Right, but PyOS_InputHook is at the C level; the only hook at the Python level is the displayhook, which doesn't get called when we need it. 

Sure, but (using tkagg as an example), the point is that Tkinter calls PyOS_InputHook for you, and by using ```draw_idle``` you can let Tkinter do the work for you.
To be specific:
* the user executes a command that causes a call to draw_idle;
* draw_idle calls _tkcanvas.after_idle(idle_draw);
* after each command, Python calls PyOS_InputHook, which points to the Tkinter event loop;
* the Tkinter event loop calls idle_draw;

which is the behavior we are trying to implement here.

>The Artist picker already accesses figure.canvas, so I thought maybe the stale setter could also call figure.canvas.draw_idle() with a try/except to catch AttributeError.

I don't see why you would need a separate ```stale``` variable. Conceptually, the ```._idle``` variable in FigureCanvasTkAgg does the same thing as the ```stale``` variable.

**Comment by WeatherGod** [comment]
I think having `stale` at the artist level is useful for providing a
backend-agnostic framework for not just redraw events, but also to allow
users to hook in their own callbacks to property updates (although traitlet
would likely do this better).

Doesn't it also help prevent unnecessary redraws of elements in the artist
tree, right?

**Comment by OceanWolf** [comment]
Apologies for the late response on this, I didn't pay it any attention as I have never got interactive mode to work (still haven't, but that's another story).

My two cents on this, I have no opinion on ``stale``, but I find the displayhook part of the implantation very awkward... it feels like we use some python technicality to get a job done that forms a fundamental part of MPL.

I mean with this solution:
```python
ln.set_color('r')
ln.set_linewidth(5)
```
only works in interactive mode...

try:
```python
import matplotlib
matplotlib.use('GTK3Agg')
from matplotlib import pyplot as plt
from gi.repository import Gtk, Gdk, GObject, GLib

fig, ax = plt.subplots()
ln, = plt.plot(range(5))

def change_props(self):
    print 'changing props'
    ln.set_lw(5)
    ln.set_color('purple')

if True:
    tbar = fig.canvas.manager.toolbar
    tbutton = Gtk.ToolButton()
    tbutton.set_label('change props')
    tbar.insert(tbutton, 0)
    tbutton.connect('clicked', change_props)
    tbar.show_all()
    
plt.show()
```

This works for me neither on 1.4.2 and on master.  Note I didn't use MEP22 (new toolbar) here because I wanted to test on 1.4.x.  With MEP22, this would become a more common occurrence with the ease of creating new tools.  This becomes even nicer as and when MEP26 (Artist Style Sheets) ever get off the ground, and perhaps we should include this as part of the MEP26 refactor as if I understand correctly that deals with precisely this problem (see https://github.com/matplotlib/matplotlib/blob/master/doc/devel/MEP/MEP26.rst).

**Comment by efiring** [comment]
@OceanWolf Try #4506.  With that, there will be no use of displayhook.

**Comment by OceanWolf** [comment]
@efiring just tried and it doesn't work, of course if I add ``draw_idle()`` after making the changes (in the function) then it works, but that defeats the point.  Anyway I think I understand it now and will make further comments over in that PR...

And about interactive mode, I think I now understand, I have to either explicitly set it as an rcParam or run ``plt.ion()``?  I guessed mpl would somehow automatically detect if I were to run it in interactive mode, but I don't see any such magic in mpl... I will make a separate PR about that.
